File: App/app/src/main/java/com/cinema/ticket_booking/ui/booking/ShowtimeAdapter.java
package com.cinema.ticket_booking.ui.booking;

import android.view.*;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import com.cinema.ticket_booking.R;
import com.cinema.ticket_booking.data.model.response.ShowtimeResponse;
import java.util.List;

public class ShowtimeAdapter extends RecyclerView.Adapter<ShowtimeAdapter.VH> {
    public interface Listener {
        void onClick(ShowtimeResponse s);
    }

    private final List<ShowtimeResponse> items;
    private final Listener listener;

    public ShowtimeAdapter(List<ShowtimeResponse> items, Listener listener) {
        this.items = items;
        this.listener = listener;
    }

    @NonNull
    @Override
    public VH onCreateViewHolder(@NonNull ViewGroup p, int t) {
        View v = LayoutInflater.from(p.getContext()).inflate(R.layout.item_showtime, p, false);
        return new VH(v);
    }

    @Override
    public void onBindViewHolder(@NonNull VH h, int pos) {
        h.bind(items.get(pos));
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    class VH extends RecyclerView.ViewHolder {
        TextView tvTime, tvType, tvPrice, tvSeats;

        VH(View v) {
            super(v);
            tvTime = v.findViewById(R.id.tvTime);
            tvType = v.findViewById(R.id.tvType);
            tvPrice = v.findViewById(R.id.tvPrice);
            tvSeats = v.findViewById(R.id.tvSeats);
        }

        void bind(ShowtimeResponse s) {
            String time = s.getStartTime();
            if (time != null && time.length() >= 16)
                time = time.substring(11, 16);
            tvTime.setText(time);
            tvType.setText(s.getScreenType());
            tvPrice.setText(String.format("%,.0fđ", s.getBasePrice()));
            tvSeats.setText(s.getAvailableSeats() + " ghế trống");
            itemView.setAlpha("CANCELLED".equals(s.getStatus()) ? 0.4f : 1f);
            itemView.setEnabled(!"CANCELLED".equals(s.getStatus()));
            itemView.setOnClickListener(v -> listener.onClick(s));
        }
    }
}

File: App/app/src/main/java/com/cinema/ticket_booking/ui/booking/ShowtimeAdapter.kt
package com.cinema.ticket_booking.ui.booking

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.cinema.ticket_booking.R
import com.cinema.ticket_booking.data.model.response.ShowtimeResponse

class ShowtimeAdapter(
    private val items: List<ShowtimeResponse>,
    private val listener: (ShowtimeResponse) -> Unit
) : RecyclerView.Adapter<ShowtimeAdapter.VH>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VH {
        val v = LayoutInflater.from(parent.context).inflate(R.layout.item_showtime, parent, false)
        return VH(v)
    }

    override fun onBindViewHolder(holder: VH, position: Int) {
        holder.bind(items[position])
    }

    override fun getItemCount(): Int = items.size

    inner class VH(v: View) : RecyclerView.ViewHolder(v) {
        private val tvTime: TextView = v.findViewById(R.id.tvTime)
        private val tvType: TextView = v.findViewById(R.id.tvType)
        private val tvPrice: TextView = v.findViewById(R.id.tvPrice)
        private val tvSeats: TextView = v.findViewById(R.id.tvSeats)

        fun bind(s: ShowtimeResponse) {
            var time = s.startTime
            if (time != null && time.length >= 16) {
                time = time.substring(11, 16)
            }
            tvTime.text = time
            tvType.text = s.screenType
            tvType.visibility = View.VISIBLE
            tvPrice.text = String.format("%,.0fđ", s.basePrice)
            tvPrice.visibility = View.VISIBLE
            tvSeats.text = "${s.availableSeats} ghế trống"
            tvSeats.visibility = View.VISIBLE
            itemView.alpha = if ("CANCELLED" == s.status) 0.4f else 1.0f
            itemView.isEnabled = "CANCELLED" != s.status
            itemView.setOnClickListener { listener(s) }
        }
    }
}
